cmd/broker-probe/main.go

```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"log"
+	"os"
+	"time"
+
+	"pack.ag/amqp"
+)
+
+func main() {
+	probeAddress := os.Getenv("PROBE_ADDRESS")
+	if probeAddress == "" {
+		log.Fatal("Error looking up PROBE_ADDRESS")
+	}
+	log.Println("Found probe address, connecting")
+
+	username := os.Getenv("PROBE_USERNAME")
+	if username == "" {
+		username = "probe"
+	}
+
+	password := os.Getenv("PROBE_PASSWORD")
+	if password == "" {
+		password = "probe"
+	}
+
+	probeTimeoutStr := os.Getenv("PROBE_TIMEOUT")
+	if probeTimeoutStr == "" {
+		probeTimeoutStr = "2s"
+	}
+
+	probeTimeout, err := time.ParseDuration(probeTimeoutStr)
+	if err != nil {
+		log.Fatal(err, "Error parsing duration")
+	}
+
+	client, err := amqp.Dial("amqp://127.0.0.1:5672", amqp.ConnSASLPlain(username, password), amqp.ConnConnectTimeout(probeTimeout), amqp.ConnProperty("product", "broker-probe"))
+	if err != nil {
+		log.Fatal(err, "Error dialing broker")
+	}
+	log.Println("Connected, creating session")
+
+	defer func() {
+		_ = client.Close()
+	}()
+
```

Comment:
We should wrap the session/link code with a https://golang.org/pkg/context/#WithTimeout.  This serves to minimise the risk that Probe invocations could stack up (adding to a storm) if things start to go wrong (say Artemis in GC death spiral).

Reply:
Mmm.  I didn't notice they didn't take a cxt.  One way I suppose would be to schedule a close of the connection after n seconds.    Looking at the link code and the session is looks like if the session/connection is closed, it will stop awaiting the response.

btw.  I have a simple Schedule implementation in pkg/consolegraphql/server/utils.go:14.  I guess you'll need to use a` once()` to organise for the connection to be closed exactly one time.

Reply:
I've attempted an approach where I spin off the probe in a goroutine and use a timeout on that in the main function.